server/src/main/java/org/elasticsearch/cluster/metadata/InternalDesiredNodesMembershipService.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class InternalDesiredNodesMembershipService implements ClusterStateListener, DesiredNodesMembershipService {
+    private final Set<DesiredNode> members;
+    private String latestHistoryId = null;
+
+    InternalDesiredNodesMembershipService() {
+        this.members = new HashSet<>();
+    }
+
+    public static InternalDesiredNodesMembershipService create(ClusterService clusterService) {
+        var tracker = new InternalDesiredNodesMembershipService();
+        clusterService.addListener(tracker);
+        return tracker;
+    }
+
+    @Override
+    public synchronized void clusterChanged(ClusterChangedEvent event) {
```

Comment:
My understanding is that `clusterChanged` events are along the same lines as cluster state operations, in that they are serialized to a single thread. That would make the `synchronized` keyword unnecessary here. Am I correct in my understanding?

Reply:
I used this approach for simplicity, but it's true that if we end up relying on this more it can become costly. I will move the `members` map to a `volatile` field to avoid such synchronization.